### ISSUE TITLE
iTerm/Safari-style tabs

### ISSUE THREAD

**Issue by y-mx-b** [wez/wezterm#1914]
**Is your feature request related to a problem? Please describe.**
Not much of a problem, but I think it'd be cool if Wezterm could support iTerm/Safari-style tabs. 

iTerm:
<img width="791" alt="Screen Shot 2022-04-24 at 7 16 13 PM" src="https://user-images.githubusercontent.com/48655352/165000740-eaf90d9e-e06a-4838-9f15-e2010b087a4c.png">

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by wez** [comment]
https://github.com/wez/wezterm/discussions/1834 is another request for this.

**Comment by digitallyserviced** [comment]
@y-mx-b The screenshot is quite simple. Is that all you are looking for it to be like? Simple tab with the name and an X to close? I think iTerm does tab tinting based on some options or rules but probably wont be a first implementation feature. 

The biggest issue I see with trying to pull this off is the fact that the tabs are so damn customizable in WezTerm. There are quite a few things that would have to be removed and assumed when making the tab widths stretch/responsive.

- Probably would force fancy tabs
- Reduce customization levels
- Disable window-right-status from being drawn into the same row/container or just disable it

**Comment by wez** [comment]
FWIW, if you're using the retro tabs, you can set https://wezfurlong.org/wezterm/config/lua/config/tab_max_width.html to a large number.

**Comment by wez** [comment]
In terms of implementation, if you wanted to take a crack at it, the relevant section of code for fancy tab bars is:

https://github.com/wez/wezterm/blob/3dfe75362ae74729e8b02f8a248df536236d5252/wezterm-gui/src/termwindow/render.rs#L681-L690

I think the thing to do is add a `min_width` constraint to the box model code that gets handled somewhere around here:

https://github.com/wez/wezterm/blob/3dfe75362ae74729e8b02f8a248df536236d5252/wezterm-gui/src/termwindow/box_model.rs#L592-L597

With that in place, then the remaining work is to add a config option for this, and then compute a reasonable min-width constraint when it is set.

**Comment by y-mx-b** [comment]
> @y-mx-b The screenshot is quite simple. Is that all you are looking for it to be like? Simple tab with the name and an X to close? I think iTerm does tab tinting based on some options or rules but probably wont be a first implementation feature. 
> 
> 
> 
> The biggest issue I see with trying to pull this off is the fact that the tabs are so damn customizable in WezTerm. There are quite a few things that would have to be removed and assumed when making the tab widths stretch/responsive.
> 
> 
> 
> - Probably would force fancy tabs
> 
> - Reduce customization levels
> 
> - Disable window-right-status from being drawn into the same row/container or just disable it
> 
> 

Yes, I mostly just want the stretched tabs. Wezterm's tab bar is incredibly customizable, but I didn't look too much into it because I wasn't very interested in customizing the tabs. I do like the look of stretched tabs that fill out the tab bat though, which is why I'd like to see this feature added to this project. I also like how the fancy tab bar looks more than the retro tab bar.

**Comment by DUOLabs333** [comment]
> FWIW, if you're using the retro tabs, you can set https://wezfurlong.org/wezterm/config/lua/config/tab_max_width.html to a large number.

This doesn't work, at least not the way expected. Is there a way to manually specify the width of new tabs?

**Comment by peterjaric** [comment]
> > FWIW, if you're using the retro tabs, you can set https://wezfurlong.org/wezterm/config/lua/config/tab_max_width.html to a large number.
> 
> This doesn't work, at least not the way expected. Is there a way to manually specify the width of new tabs?

This is my experience too. I went into the settings and noticed that I already had specified this (set to 50). If I set config.tab_max_width to 1, my tabs get really small, but at some point (16?) the setting does not have any effect. So if I set it to for example 1000 my tabs still do not fill out all the space of the window:

![image](https://github.com/user-attachments/assets/0a380a2f-2427-4fdc-8096-7d5765dc128e)

**Comment by ogarcia** [comment]
> FWIW, if you're using the retro tabs, you can set https://wezfurlong.org/wezterm/config/lua/config/tab_max_width.html to a large number.

@wez The only drawback is that the width of the tab always depends on the text it contains, so even if you put a very high value if the text is _top_ (because you are running `top`) then the tab occupies very little:

![Tabs](https://github.com/user-attachments/assets/8634484e-3483-4226-89c4-314221a3ec79)

As I see it, the objective is that the tabs always occupy all the possible width and that they all have the same size regardless of the content. Something like this:
1. We get the width of the window.
2. We subtract from it what occupies the _add tab button_ that we always have on the right.
3. What we have left is the total width we have for the tabs. From here that width we divide it by the number of tabs that we have and that is the width of the tabs (and they are all of the same width). 
4. TL/DR: The formula: `(WIDTH_OF_WINDOW - WIDTH_PLUS_SIGN) / NUMBER_OF_TABS = WIDTH_OF_TABS`.

In my case it is a functionality that I would love to have in the retro tabs.
